lib/fs/fakefs_test.go

```diff
@@ -172,3 +178,722 @@ func TestFakeFSRead(t *testing.T) {
+			t.Run(name, func(t *testing.T) {
+				test.impl(t, filesystem.fs)
+				if err := cleanup(filesystem.fs); err != nil {
+					t.Errorf("cleanup failed: %s", err)
+				}
+			})
+		}
+	}
+}
+
+func testFakeFSCaseInsensitive(t *testing.T, fs Filesystem) {
+	bs1 := []byte("test")
+
+	err := fs.Mkdir("/fUbar", 0755)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	fd1, err := fs.Create("fuBAR/SISYPHOS")
+	if err != nil {
+		t.Fatalf("could not create file: %s", err)
+	}
+
+	_, err = fd1.Write(bs1)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	fd1.Close()
+
+	// Try reading from the same file with different filenames
+	fd2, err := fs.Open("Fubar/Sisyphos")
+	if err != nil {
+		t.Fatalf("could not open file by its case-differing filename: %s", err)
+	}
+
+	if _, err := fd2.Seek(0, io.SeekStart); err != nil {
+		t.Fatal(err)
+	}
+
+	bs2, err := ioutil.ReadAll(fd2)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	fd2.Close()
+
+	if len(bs1) != len(bs2) {
+		t.Errorf("wrong number of bytes, expected %d, got %d", len(bs1), len(bs2))
+	}
+}
+
+func testFakeFSCaseInsensitiveMkdirAll(t *testing.T, fs Filesystem) {
+	err := fs.MkdirAll("/fOO/Bar/bAz", 0755)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	fd, err := fs.OpenFile("/foo/BaR/BaZ/tESt", os.O_CREATE, 0644)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if err = fd.Close(); err != nil {
+		t.Fatal(err)
+	}
+
+	if err = fs.Rename("/FOO/BAR/baz/tesT", "/foo/baR/BAZ/Qux"); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestFakeFSDirNames(t *testing.T) {
+	fs := newFakeFilesystem("/fbr")
+	testDirNames(t, fs)
+}
+
+func testFakeFSDirNamesInsens(t *testing.T, fs Filesystem) {
+	testDirNames(t, fs)
+}
+
+func testDirNames(t *testing.T, fs Filesystem) {
+	t.Helper()
+	filenames := []string{"fOO", "Bar", "baz"}
+	for _, filename := range filenames {
+		fd, err := fs.Create("/" + filename)
+		if err != nil {
+			t.Errorf("Could not create %s: %s", filename, err)
+		}
+		fd.Close()
+	}
+
+	assertDir(t, fs, "/", filenames)
+}
+
+func assertDir(t *testing.T, fs Filesystem, directory string, filenames []string) {
+	t.Helper()
+	got, err := fs.DirNames(directory)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if path.Clean(directory) == "/" {
+		filenames = append(filenames, ".stfolder")
+	}
+	sort.Strings(filenames)
+	sort.Strings(got)
+
+	if !reflect.DeepEqual(got, filenames) {
+		t.Errorf("want %s, got %s", filenames, got)
+	}
+}
+
+func testFakeFSStatInsens(t *testing.T, fs Filesystem) {
+	if err := fs.Mkdir("/foo", 0755); err != nil {
+		t.Fatal(err)
+	}
+
+	fd, err := fs.Create("/Foo/aaa")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	fd.Close()
+
+	info, err := fs.Stat("/FOO/AAA")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if _, err = fs.Stat("/fOO/aAa"); err != nil {
+		t.Fatal(err)
+	}
+
+	if info.Name() != "AAA" {
+		t.Errorf("want AAA, got %s", info.Name())
+	}
+
+	fd1, err := fs.Open("/FOO/AAA")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if info, err = fd1.Stat(); err != nil {
+		t.Fatal(err)
+	}
```

Comment:
This doesn't check anything.

Reply:
Yep, Open can be omitted, but it basically checks that fd.Stat doesn't spoil the name as well as fs.Stat.